File: framework/languages/node/e2e/AutomaticTurnDispatch/Server/Session/session-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import {
  ZLinkMessageFlowLogMode,
  type ZLinkEntrySpot,
  type ZLinkEntrySpotContext,
  type ZLinkMessage,
  type ZLinkSpotActorJoinResult
} from '@zlink-systems/framework';
import { ZLinkRedisLocationStore } from '@zlink-systems/framework-locations-redis';
import { ZLinkModule, zlinkFramework } from '@zlink-systems/nestjs';
import { AutomaticTurnDispatchNames } from '../../Shared/messages';
import { EvidenceStore } from './Support/evidence-store';
import { closeHttpServer, startHttpServer } from './Support/http-server';
import { createAutomaticTurnConfigurationModule } from '../../configuration';
import { SESSION_OPTIONS, validateSessionOptions } from './Configuration/session-options';
import type { SessionOptions } from './Configuration/session-options';
import { AwaitSessionFactory } from './Handlers/await-session';

class AwaitSessionEntrySpot implements ZLinkEntrySpot {
  readonly context!: ZLinkEntrySpotContext;

  async onActorJoin(actorId: string, request: ZLinkMessage): Promise<ZLinkSpotActorJoinResult> {
    void actorId;
    void request;
    return { accepted: true };
  }

  async onJoinedActor(actor: import('@zlink-systems/framework').ZLinkActor): Promise<void> { void actor; }

  async onLeaveActor(actor: import('@zlink-systems/framework').ZLinkActor): Promise<void> { void actor; }

  async onDisconnectActor(actor: import('@zlink-systems/framework').ZLinkActor): Promise<void> { void actor; }
}

export async function startSessionHost(): Promise<void> {
  let stopping = false;
  const configuration = createAutomaticTurnConfigurationModule(SESSION_OPTIONS, validateSessionOptions);

  class SessionModule {}
  Module({
    imports: [
      configuration,
      ZLinkModule.forRootFactory({
        imports: [configuration],
        inject: [SESSION_OPTIONS],
        useFactory: (value: unknown) => {
          const options = value as SessionOptions;
          fs.mkdirSync(options.logDir, { recursive: true });
          const locationStore = new ZLinkRedisLocationStore({
            url: `redis://${options.redisEndpoint}`,
            keyPrefix: options.redisKeyPrefix
          });
          const builder = zlinkFramework();
          builder
            .addLocationStore(locationStore)
            .configureDispatch()
              .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
              .traceLogFile(`${options.logDir}/${options.rid}-flow.log`)
              .traceLabel(options.rid);

          const controlMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.controlChannel)
            .listen(options.controlRouterEndpoint)
            .routingId(options.rid);
          controlMesh.channel(AutomaticTurnDispatchNames.controlChannel).server();
          for (const endpoint of options.playControlEndpoints) controlMesh.peerConnections().connect(endpoint);
          const routeMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.spotRouteChannel)
            .listen(options.spotRouteEndpoint)
            .routingId(options.rid);
          routeMesh.channel(AutomaticTurnDispatchNames.spotRouteChannel).server();
          for (const endpoint of options.playSpotRouteEndpoints) routeMesh.peerConnections().connect(endpoint);
          const spotMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.spotChannel)
            .routingId(options.rid)
            .listen(options.spotRouterEndpoint);
          spotMesh.objects().server().addEntrySpot(AwaitSessionEntrySpot);
          spotMesh.channel(AutomaticTurnDispatchNames.spotChannel).server();
          for (const peer of options.spotRouterPeers) spotMesh.peerConnections().connect(peer.rid, peer.endpoint);
          builder.addStreamNode(AutomaticTurnDispatchNames.streamNode)
            .bind(options.streamEndpoint)
            .registerSession(AwaitSessionFactory);

          return builder.build();
        }
      })
    ],
    providers: [
      {
        provide: EvidenceStore,
        inject: [SESSION_OPTIONS],
        useFactory: (options: SessionOptions) => new EvidenceStore(options.rid, options.evidenceFile)
      },
      AwaitSessionEntrySpot,
      AwaitSessionFactory
    ]
  })(SessionModule);

  const app = await NestFactory.createApplicationContext(SessionModule, { logger: false, abortOnError: false });
  const options = app.get<SessionOptions>(SESSION_OPTIONS);
  const evidence = app.get(EvidenceStore);
  const server = await startHttpServer(options.httpUrl, [
    { method: 'GET', path: '/health', handle: () => ({ status: 'ready', role: 'session', rid: options.rid }) },
    { method: 'GET', path: '/evidence', handle: () => evidence.snapshot() },
    {
      method: 'POST',
      path: '/shutdown',
      handle: () => {
        stopping = true;
        return { status: 'stopping' };
      }
    }
  ]);
  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}

File: framework/languages/dotnet/e2e/AutomaticTurnDispatch/Server/Session/SessionHostFactory.cs
using Microsoft.Extensions.Configuration;

using Systems.Zlink;
using AutomaticTurnDispatch.Server.Session.Support;
using AutomaticTurnDispatch.Shared;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Locations.Redis;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Configuration;
using Zlink.Framework.Contracts.Dispatch;
using Zlink.Framework.Contracts.Errors;
using Zlink.Framework.Contracts.Actors;
using SessionServerOptions = AutomaticTurnDispatch.Server.Session.Support.SessionOptions;
using AwaitStreamSession = AutomaticTurnDispatch.Server.Session.Support.AwaitSession;

namespace AutomaticTurnDispatch.Server.Session;

internal static class SessionHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = SessionServerOptions.Parse(args);
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Configuration.Sources.Clear();
        builder.Configuration.AddInMemoryCollection();
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        builder.Services.AddSingleton(new EvidenceStore(options.Rid, options.EvidenceFile));
        builder.Services.AddSingleton(new NodeOptions(options.Rid));
        builder.Services.AddZLinkFramework(framework =>
        {
            framework.AddLocationStore(new ZLinkRedisLocationStore(redis => redis
                .SetConnectionString(options.RedisEndpoint)
                .SetKeyPrefix(options.RedisKeyPrefix)));
            framework.AddRelocationStore(new ZLinkRedisRelocationStore(redis => redis
                .SetConnectionString(options.RedisEndpoint)
                .SetKeyPrefix($"{options.RedisKeyPrefix}:relocation")));
            framework.AddHandlersFromAssemblyOf(typeof(Program));
            framework.ConfigureDispatch()
                .MessageFlow(ZLinkRuntimeMessageFlowMode.KeyTransitions)
                .TraceLogFile(Path.Combine(options.LogDir, $"{options.Rid}-flow.log"))
                .TraceLabel(options.Rid);
            var controlMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.ControlChannel)
                .Listen(options.ControlEndpoint)
                .SetRoutingIdPrefix(options.Rid);
            controlMesh.Channel(AutomaticTurnDispatchNames.ControlChannel).Client();
            var spotRouteMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.SpotRouteChannel)
                .Listen("tcp://127.0.0.1:0")
                .SetRoutingIdPrefix(options.Rid);
            spotRouteMesh.Channel(AutomaticTurnDispatchNames.SpotRouteChannel).Server().SetWeight(0);
            var mesh23 = framework.AddRouteMesh(AutomaticTurnDispatchNames.SpotChannel)
                .Listen(options.SpotRouterEndpoint)
                .SetRoutingIdPrefix(options.Rid);
            mesh23.Objects().Server()
                .AddEntrySpot<SessionAwaitEntrySpot>()
                .AddActorFactory<SessionAwaitActor, SessionAwaitActorFactory>(
                    AutomaticTurnDispatchNames.ActorType,
                    options: null,
                    ZLinkRelocationPolicy<SessionAwaitActor>.Recreate);
            mesh23.Channel(AutomaticTurnDispatchNames.SpotChannel).Server();
            framework.AddStreamNode(AutomaticTurnDispatchNames.StreamNode)
                .Bind(options.StreamEndpoint)
                .EnableActorDispatch()
                .AddSession<AwaitStreamSession>();
        });

        var app = builder.Build();
        app.MapGet("/health", () => Results.Ok(new { status = "ready", role = "session", options.Rid }));
        app.MapGet("/topology/ready", async (
            string meshName,
            string rid,
            IZLinkRouteMeshRuntime runtime,
            IZLinkRouteClient routes,
            CancellationToken cancellationToken) =>
        {
            var snapshot = runtime.Snapshot(meshName);
            var snapshotReady = snapshot.Peers.Any(peer =>
                peer.Ready
                && (string.Equals(
                        peer.Rid.ToString(),
                        rid,
                        StringComparison.Ordinal)
                    || peer.Rid.ToString().StartsWith(
                        $"{rid}-",
                        StringComparison.Ordinal)));
            if (!snapshotReady
                || !string.Equals(
                    meshName,
                    AutomaticTurnDispatchNames.ControlChannel,
                    StringComparison.Ordinal))
                return Results.Ok(new { ready = snapshotReady });

            try
            {
                var target = snapshot.Peers
                    .Where(peer => peer.Ready)
                    .Select(peer => peer.Rid)
                    .First(peer => peer.ToString().StartsWith(
                        $"{rid}-",
                        StringComparison.Ordinal));
                _ = await routes.RequestToNode(
                        meshName,
                        target,
                        new AwaitEvidenceReq($"readiness-{rid}"))
                    .Timeout(TimeSpan.FromMilliseconds(500))
                    .Async<AwaitEvidenceRes>(cancellationToken);
                return Results.Ok(new { ready = true });
            }
            catch (Exception error) when (
                error is ZLinkFrameworkException
                    or TimeoutException
                    or OperationCanceledException)
            {
                return Results.Ok(new { ready = false });
            }
        });
        app.MapGet("/evidence", (EvidenceStore evidence) => Results.Ok(evidence.Snapshot()));
        return app;
    }
}
